src/pybind/mgr/dashboard/services/auth/auth.py

```diff
@@ -25,6 +30,74 @@
 })
 
 
+class AuthType(str, Enum):
+    LOCAL = 'local'
+    SAML2 = 'saml2'
+    OAUTH2 = 'oauth2'
+
+
+class BaseAuth(abc.ABC):
+    LOGIN_URL: str
+    LOGOUT_URL: str
+    sso: bool
+
+    @staticmethod
+    def from_protocol(protocol: AuthType) -> Type["BaseAuth"]:
+        for subclass in BaseAuth.__subclasses__():
+            if subclass.__name__.lower() == protocol:
+                return subclass
+            for subsubclass in subclass.__subclasses__():
+                if subsubclass.__name__.lower() == protocol:
+                    return subsubclass
+        raise ValueError(f"Unknown auth backend: '{protocol}'")
+
+    @classmethod
+    def from_db(cls, db: Optional['SsoDB'] = None) -> Type["BaseAuth"]:
+        if db is None:
+            protocol = mgr.SSO_DB.protocol
+        else:
+            protocol = db.protocol
+        return cls.from_protocol(protocol)
+
+    class Config(TypedDict):  # pylint: disable=inherit-non-class
+        pass
+
+    @abc.abstractmethod
+    def to_dict(self) -> 'Config':
+        pass
+
+    @classmethod
+    @abc.abstractmethod
+    def from_dict(cls, s_dict) -> 'BaseAuth':
+        pass
+
+    @classmethod
+    def get_auth_name(cls):
+        return cls.__name__.lower()
+
+
+class Local(BaseAuth):
+    LOGIN_URL = '#/login'
+    LOGOUT_URL = '#/login'
+    sso = False
+
+    @classmethod
+    def get_auth_name(cls):
+        return cls.__name__.lower()
+
+    def to_dict(self) -> 'BaseAuth.Config':
+        return BaseAuth.Config()
```

Comment:
> This is not using self, hence it's not an instance method but a classmethod. It's returning a default empty object.

My bad, I was previously using self but I was hitting a typeError with the return type, and after changing it I missed this.
It's not using the class so it would be a `staticmethod` then, right?

However, having said that, the reason why I wanted `to_dict` to be an instance method was to access the class instance values as was [being done in the Saml2](https://github.com/ceph/ceph/blob/c37a00af2113d089d40b9d6df347fcb940efa41a/src/pybind/mgr/dashboard/services/auth/saml2.py#L14).

If I change this to a `staticmethod` that would mean I need to change the way of storting the auth config values, so instead of doing it at the object level, should I do it at the class level?

Reply:
That's my point. A `to_dict()` method is expected to "serialize" the properties of a class **instance** as a dict, but in this case it does nothing with `self` and instead returns a generic dict from another base class.